src/ethereum/starknet/state_root.rs

```diff
@@ -0,0 +1,294 @@
+    ///
+    /// ## Dealing with Infura's cap error
+    ///
+    /// Infura responds with an error if the query is too large, and would
+    /// result in more than 10 000 logs. To deal with this, we use a binary
+    /// search like query range. We need at least 2 logs (1 is the last known)
+    /// to have a new update.
+    async fn retrieve_logs(
+        &mut self,
+    ) -> Result<VecDeque<LogWithOrigin<StateUpdateLog>>, GetRootError> {
+        let from_block = self
+            .last
+            .as_ref()
+            .map(|update| update.origin.block_number)
+            .unwrap_or_default();
+
+        // The largest stride we are allowed to take. This gets
+        // set if we encounter the Infura result cap error.
+        //
+        // Allows us to perform a binary search of the query range space.
+        let mut stride_cap = None;
+
+        loop {
+            let to_block = from_block.saturating_add(self.stride);
+
+            let mut logs = match self.get_logs(from_block, to_block).await {
+                Ok(logs) => logs,
+                Err(GetLogsError::QueryLimit) => {
+                    // Infura result cap error. We need to reduce our query range.
+                    stride_cap = Some(self.stride);
+                    self.stride = (self.stride / 2).max(1);
+
+                    continue;
+                }
+                Err(GetLogsError::Other(other)) => return Err(GetRootError::Other(other)),
+            };
+
+            // Check for reorgs. Only required if there was a last known update to validate.
+            if let Some(last) = self.last.as_ref() {
+                // We queried for logs starting from the same block as last known. We need to account
+                // for logs that occurred in the same block and transaction, but with a smaller log index.
+                //
+                // If the last known log is not in the set then we have a reorg event.
+                loop {
+                    match logs.pop_front() {
+                        // Log from the same origin point, but smaller log index.
+                        Some(log)
+                            if log.origin == last.origin && log.log_index < last.log_index =>
+                        {
+                            continue
+                        }
+                        // Log found. Yay!
+                        Some(log) if &log == last => break,
+                        _ => return Err(GetRootError::Reorg),
+                    }
+                }
+            }
+
+            // At this point we have validated and removed the last known log.
+            // If there are no new logs then either:
+            //  1. there are no new logs on L1, or
+            //  2. we need to increase our stride
+            if logs.is_empty() {
+                let latest_on_chain = self.websocket.eth().block_number().await?.as_u64();
+
+                if to_block <= latest_on_chain {
```

Comment:
I've just noticed this is incorrect.
```suggestion
                if to_block < latest_on_chain {
```